README.md
Holds APIRepository, which is primarily used for making requests to a server

## Getting Started

TODO:


CHANGELOG.md
## 0.3.0

- Upgraded event_db to 0.2.0

## 0.2.2

- Added convenience function for loading bodies in Responses as Lists of a GenericModel

## 0.2.1

- Updated http to ^1.0.0

## 0.2.0

- Added MultipartRequests to requirement for APIRequesters

## 0.1.2

- Added logout option to APIRepository

## 0.1.1

- Added ApiDatabaseRepository - an implementation of DatabaseRepository that is backed by an APIRequester
- Added StubAPIRequester - an implementation of APIRequester that will always return the same response regardless of urlSuffix

## 0.1.0+1

- Fixed exports to export all the libraries

## 0.1.0

- Initial version.


test/fake_test.dart
import 'package:event_api/event_api.dart';
import 'package:event_authentication/event_authentication.dart';
import 'package:event_db/event_db.dart';
import 'package:http/http.dart';
import 'package:test/test.dart';

void main() {
  group('FakeAPIRequester', () {
    test('Fail', () async {
      final database =
          FakeDatabaseRepository(constructors: {EncodedJWT: EncodedJWT.new});
      final requester = FakeAPIRequester(
        fakeRequestMap: {},
      );

      final repository = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final response =
          await repository.request('POST', 'login', (request) => null);

      expect(response.statusCode, 500);
    });
    test('Login', () async {
      final database =
          FakeDatabaseRepository(constructors: {EncodedJWT: EncodedJWT.new});
      final requester = FakeAPIRequester(
        fakeRequestMap: {
          'login': (request) async => StreamedResponse(
                Stream.value([]),
                200,
                headers: {}..authorization = await JWTSigner(
                    () => 'cool',
                    issuer: 'example.com',
                  ).createToken(BaseJWT()..expiry = const Duration(hours: 5)),
              ),
          'loggedIn': (request) async => StreamedResponse(
                Stream.value(request.headers.authorization!.codeUnits),
                200,
              )
        },
      );

      final repository = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final response =
          await repository.request('POST', 'login', (request) => null);

      expect(response.statusCode, 200);
      final jwt = await repository.jwt;
      expect(jwt!.jwt.expiry, const Duration(hours: 5));

      final loggedInResponse =
          await repository.request('POST', 'loggedIn', (request) => null);

      final loggedInData = await loggedInResponse.body;
      final returnedData = BaseJWT.fromToken(loggedInData);

      expect(returnedData.expiry, const Duration(hours: 5));

      await repository.logout();
      expect(await repository.jwt, null);
    });
  });
}


test/server_test.dart
import 'package:event_api/event_api.dart';
import 'package:event_authentication/event_authentication.dart';
import 'package:event_db/event_db.dart';
import 'package:test/test.dart';

void main() {
  group('ServerAPIRequester', () {
    test('Fail', () async {
      final database =
          FakeDatabaseRepository(constructors: {EncodedJWT: EncodedJWT.new});
      final requester = ServerAPIRequester(
        apiServer: 'https://test.vhcblade.com/',
        website: 'https://test.vhcblade.com/',
      );

      final repository = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final response =
          await repository.request('POST', 'amazing', (request) => null);

      expect(response.statusCode, 404);
      expect(await response.body, 'Route not found');
    });
    test('Login', () async {
      final database =
          FakeDatabaseRepository(constructors: {EncodedJWT: EncodedJWT.new});
      final requester = ServerAPIRequester(
        apiServer: 'https://test.vhcblade.com/',
        website: 'https://test.vhcblade.com/',
      );

      final repository = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final response = await repository.request(
        'POST',
        'login/email',
        (request) => request.body = (EmailLoginRequest()
              ..email = 'example@example.com'
              ..password = 'example')
            .toJsonString(),
      );

      expect(response.statusCode, 200);
      final jwt = await repository.jwt;
      expect(jwt!.jwt.expiry, const Duration(hours: 2));

      await repository.request(
        'POST',
        'authenticated',
        (request) => null,
      );

      final newJwt = await repository.jwt;
      expect(newJwt!.jwt.expiry, const Duration(hours: 2));
      expect(
        newJwt.jwt.dateIssued.microsecondsSinceEpoch,
        isNot(jwt.jwt.dateIssued.microsecondsSinceEpoch),
      );
    });
    test('No Server', () async {
      final database =
          FakeDatabaseRepository(constructors: {EncodedJWT: EncodedJWT.new});
      final requester = ServerAPIRequester(
        apiServer: 'https://imaginary.vhcblade.com/',
        website: 'https://imaginary.vhcblade.com/',
      );

      final repository = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final response = await repository.request(
        'POST',
        'login/email',
        (request) => request.body = (EmailLoginRequest()
              ..email = 'example@example.com'
              ..password = 'example')
            .toJsonString(),
      );

      expect(response.statusCode, 504);
    });
  });
}


test/database_test.dart
import 'dart:convert';

import 'package:event_api/event_api.dart';
import 'package:event_authentication/event_authentication.dart';
import 'package:event_db/event_db.dart';
import 'package:http/http.dart';
import 'package:test/test.dart';

void main() {
  group('ApiDatabaseRepository', () {
    test('Missing Constructor', () async {
      final database =
          FakeDatabaseRepository(constructors: {EncodedJWT: EncodedJWT.new});
      final requester = createUnreachableRequester();

      final api = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final apiDatabase = ApiDatabaseRepository(
        apiRepository: api,
        baseURL: 'unreachable',
        constructors: {EncodedJWT: EncodedJWT.new},
      );
      expect(
        () => apiDatabase.getInstance<EmailLoginRequest>(),
        throwsArgumentError,
      );
    });
    test('Unreachable', () async {
      final database =
          FakeDatabaseRepository(constructors: {EncodedJWT: EncodedJWT.new});
      final requester = createUnreachableRequester();

      final api = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final apiDatabase = ApiDatabaseRepository(
        apiRepository: api,
        baseURL: 'unreachable',
        constructors: {EmailLoginRequest: EmailLoginRequest.new},
      );
      var i = 0;
      apiDatabase.errorStream.stream.listen((_) => i++);

      expect(
        (await apiDatabase.saveModel('email', createRequest())).toMap(),
        createRequest().toMap(),
      );
      await Future<void>.delayed(Duration.zero);
      expect(i, 1);

      expect(
        (await apiDatabase.saveModel('email', createRequest()..id = 'Cool'))
            .toMap(),
        (createRequest()..id = 'Cool').toMap(),
      );
      await Future<void>.delayed(Duration.zero);
      expect(i, 2);

      expect((await apiDatabase.findModel('email', 'Cool'))?.toMap(), null);
      await Future<void>.delayed(Duration.zero);
      expect(i, 3);

      expect(
        await apiDatabase.findAllModelsOfType('email', EmailLoginRequest.new),
        <EmailLoginRequest>[],
      );
      await Future<void>.delayed(Duration.zero);
      expect(i, 4);

      expect(
        await apiDatabase.deleteModel('email', createRequest()..id = 'Cool'),
        false,
      );
      await Future<void>.delayed(Duration.zero);
      expect(i, 5);
    });
    test('Not Found', () async {
      final database =
          FakeDatabaseRepository(constructors: {EncodedJWT: EncodedJWT.new});
      final requester = createNotFoundRequester();

      final api = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final apiDatabase = ApiDatabaseRepository(
        apiRepository: api,
        baseURL: 'notFound',
        constructors: {EmailLoginRequest: EmailLoginRequest.new},
      );
      var i = 0;
      apiDatabase.errorStream.stream.listen((_) => i++);

      expect(
        (await apiDatabase.saveModel('email', createRequest())).toMap(),
        createRequest().toMap(),
      );
      await Future<void>.delayed(Duration.zero);
      expect(i, 0);

      expect(
        (await apiDatabase.saveModel('email', createRequest()..id = 'Cool'))
            .toMap(),
        (createRequest()..id = 'Cool').toMap(),
      );
      await Future<void>.delayed(Duration.zero);
      expect(i, 0);

      expect((await apiDatabase.findModel('email', 'Cool'))?.toMap(), null);
      await Future<void>.delayed(Duration.zero);
      expect(i, 0);

      expect(
        await apiDatabase.findAllModelsOfType('email', EmailLoginRequest.new),
        <EmailLoginRequest>[],
      );
      await Future<void>.delayed(Duration.zero);
      expect(i, 0);

      expect(
        await apiDatabase.deleteModel('email', createRequest()..id = 'Cool'),
        false,
      );
      await Future<void>.delayed(Duration.zero);
      expect(i, 0);
    });
    test('Expected', () async {
      final database = FakeDatabaseRepository(
        constructors: {EmailLoginRequest: EmailLoginRequest.new},
      );
      final requester = createRequester(database);

      final api = APIRepository(
        database: SpecificDatabase(database, 'jwt'),
        requester: requester,
      );

      final apiDatabase = ApiDatabaseRepository(
        apiRepository: api,
        baseURL: 'cool',
        constructors: {EmailLoginRequest: EmailLoginRequest.new},
      );
      var i = 0;
      apiDatabase.errorStream.stream.listen((_) => i++);

      expect(
        (await apiDatabase.saveModel('value', createRequest())).toMap(),
        (createRequest()..idSuffix = 'id').toMap(),
      );

      expect(
        (await apiDatabase.findModel<EmailLoginRequest>(
          'value',
          (EmailLoginRequest()..idSuffix = 'id').id!,
        ))
            ?.toMap(),
        (createRequest()..idSuffix = 'id').toMap(),
      );

      expect(
        (await apiDatabase.findAllModelsOfType('value', EmailLoginRequest.new))
            .map((e) => e.toMap())
            .toList(),
        [(createRequest()..idSuffix = 'id').toMap()],
      );

      expect(
        await apiDatabase.deleteModel(
          'value',
          createRequest()..idSuffix = 'id',
        ),
        true,
      );

      expect(
        (await apiDatabase.findModel<EmailLoginRequest>(
          'value',
          (EmailLoginRequest()..idSuffix = 'id').id!,
        ))
            ?.toMap(),
        null,
      );

      expect(
        await apiDatabase.deleteModel(
          'value',
          createRequest()..idSuffix = 'id',
        ),
        false,
      );

      await Future<void>.delayed(Duration.zero);
      expect(i, 0);
    });
  });
}

EmailLoginRequest createRequest() => EmailLoginRequest()
  ..email = 'myEmail@example.com'
  ..password = 'example';

APIRequester createRequester(DatabaseRepository repository) {
  return FakeAPIRequester(
    fakeMultipartRequestMap: {},
    fakeRequestMap: {
      'cool/value': (request) async {
        switch (request.method) {
          case 'GET':
            final models = await repository.findAllModelsOfType(
              'database',
              EmailLoginRequest.new,
            );
            final response = json.encode(models.map((e) => e.toMap()).toList());
            return StreamedResponse(Stream.value(response.codeUnits), 200);
          case 'POST':
            final model = EmailLoginRequest()..loadFromJsonString(request.body);
            model.idSuffix = 'id';
            await repository.saveModel('database', model);
            return StreamedResponse(Stream.value(model.id!.codeUnits), 200);
          default:
            return StreamedResponse(Stream.value([]), 404);
        }
      },
      'cool/value/Email Login::id': (request) async {
        switch (request.method) {
          case 'GET':
            final model = await repository.findModel<EmailLoginRequest>(
              'database',
              'Email Login::id',
            );
            if (model == null) {
              return StreamedResponse(Stream.value([]), 404);
            }
            return StreamedResponse(
              Stream.value(model.toJsonString().codeUnits),
              200,
            );
          case 'DELETE':
            final model = EmailLoginRequest()..loadFromJsonString(request.body);
            final success = await repository.deleteModel<EmailLoginRequest>(
              'database',
              model,
            );
            if (!success) {
              return StreamedResponse(Stream.value([]), 404);
            }
            return StreamedResponse(
              Stream.value(model.toJsonString().codeUnits),
              200,
            );
          case 'PUT':
            final model = EmailLoginRequest()..loadFromJsonString(request.body);
            model.idSuffix = 'id';
            await repository.saveModel('database', model);
            return StreamedResponse(Stream.value(model.id!.codeUnits), 200);
          default:
            return StreamedResponse(Stream.value([]), 404);
        }
      },
    },
  );
}

APIRequester createNotFoundRequester() {
  return StubAPIRequester(
    multipartStubResponse: (_) {
      return StreamedResponse(Stream.value([]), 404);
    },
    stubResponse: (request) async {
      return StreamedResponse(Stream.value([]), 404);
    },
  );
}

APIRequester createUnreachableRequester() {
  return StubAPIRequester(
    multipartStubResponse: (_) {
      return StreamedResponse(Stream.value([]), 404);
    },
    stubResponse: (request) async {
      return StreamedResponse(Stream.value([]), 504);
    },
  );
}


lib/src/fake.dart
import 'dart:async';

import 'package:event_api/event_api.dart';
import 'package:http/http.dart';

/// A fake version of [APIRequester]. Use this in tests or help develop your
/// APIs without having to depend on an external server.
class FakeAPIRequester implements APIRequester {
  /// [fakeRequestMap] holds the fake responses for requests.
  FakeAPIRequester({
    required this.fakeRequestMap,
    required this.fakeMultipartRequestMap,
    this.website = 'https://example.com/',
    this.apiServer = 'https://api.example.com/',
  });

  /// Holds the fake responses for requests.
  ///
  /// The key is the urlSuffix used for the request.
  final Map<String, FutureOr<StreamedResponse> Function(Request)>
      fakeRequestMap;

  /// Holds the fake responses for multipart requests.
  ///
  /// The key is the urlSuffix used for the request.
  final Map<String, FutureOr<StreamedResponse> Function(MultipartRequest)>
      fakeMultipartRequestMap;

  @override
  final String apiServer;

  @override
  final String website;

  @override
  FutureOr<StreamedResponse> request(
    String method,
    String urlSuffix,
    FutureOr<void> Function(Request) addToRequest,
  ) async {
    if (!fakeRequestMap.containsKey(urlSuffix)) {
      return StreamedResponse(Stream.value([]), 500);
    }

    final fullUrl = '$apiServer$urlSuffix';
    final request = Request(method, Uri.parse(fullUrl));

    await addToRequest(request);

    return fakeRequestMap[urlSuffix]!(request);
  }

  @override
  FutureOr<StreamedResponse> multipartRequest(
    String method,
    String urlSuffix,
    FutureOr<void> Function(MultipartRequest request) addToRequest,
  ) async {
    if (!fakeMultipartRequestMap.containsKey(urlSuffix)) {
      return StreamedResponse(Stream.value([]), 500);
    }

    final fullUrl = '$apiServer$urlSuffix';
    final request = MultipartRequest(method, Uri.parse(fullUrl));

    await addToRequest(request);

    return fakeMultipartRequestMap[urlSuffix]!(request);
  }
}

/// A stub version of [APIRequester]. Unlike [FakeAPIRequester], this will
/// always respond with the same response regardless of the urlSuffix.
class StubAPIRequester implements APIRequester {
  /// [stubResponse] is the stub response that will be returned for all requests
  /// to this reqeuster.
  StubAPIRequester({
    required this.stubResponse,
    required this.multipartStubResponse,
    this.website = 'https://example.com/',
    this.apiServer = 'https://api.example.com/',
  });

  @override
  final String apiServer;

  @override
  final String website;

  /// The stub response that will be returned for all requests
  /// to this reqeuster.
  final FutureOr<StreamedResponse> Function(Request) stubResponse;

  /// The stub response that will be returned for all multipart requests
  /// to this reqeuster.
  final FutureOr<StreamedResponse> Function(MultipartRequest)
      multipartStubResponse;

  @override
  FutureOr<StreamedResponse> request(
    String method,
    String urlSuffix,
    void Function(Request request) addToRequest,
  ) async {
    final fullUrl = '$apiServer$urlSuffix';
    final request = Request(method, Uri.parse(fullUrl));

    addToRequest(request);

    return stubResponse(request);
  }

  @override
  FutureOr<StreamedResponse> multipartRequest(
    String method,
    String urlSuffix,
    void Function(MultipartRequest request) addToRequest,
  ) {
    final fullUrl = '$apiServer$urlSuffix';
    final request = MultipartRequest(method, Uri.parse(fullUrl));

    addToRequest(request);

    return multipartStubResponse(request);
  }
}


lib/src/model.dart
import 'package:event_authentication/event_authentication.dart';
import 'package:event_db/event_db.dart';
import 'package:tuple/tuple.dart';

/// This represents a JWT that has been retrieved from a server. This helps
/// preserve the tokenized version of the JWT.
class EncodedJWT extends GenericModel {
  /// Default Constructor
  EncodedJWT();

  /// Creates a new EncodedJWT with its [token] value set to [token].
  factory EncodedJWT.fromToken(String token) => EncodedJWT()..token = token;

  /// Converts [token] into a [BaseJWT] for easier use.
  BaseJWT get jwt => BaseJWT.fromToken(token);

  /// The token that represents the jwt.
  late String token;

  @override
  Map<String, Tuple2<Getter<dynamic>, Setter<dynamic>>> getGetterSetterMap() =>
      {
        'token': Tuple2(() => token, (val) => token = '$val'),
      };

  @override
  String get type => 'EncodedJWT';
}


lib/src/server.dart
import 'dart:async';
import 'dart:convert';

import 'package:event_api/event_api.dart';
import 'package:http/http.dart';

/// IMplementation of [APIRequester] that sends actual requests to an
/// external server.
class ServerAPIRequester implements APIRequester {
  /// [apiServer] is the server to send requests to.
  ///
  /// [website] is the expected client-side website to be used as the origin
  ServerAPIRequester({required this.apiServer, required this.website});
  @override
  final String apiServer;
  @override
  final String website;

  /// The client used to send requests to the server.
  late final client = Client();

  @override
  Future<StreamedResponse> request(
    String method,
    String urlSuffix,
    FutureOr<void> Function(Request) addToRequest,
  ) async {
    final fullUrl = '$apiServer$urlSuffix';
    final request = Request(method, Uri.parse(fullUrl));
    await addToRequest(request);
    try {
      return await client.send(request);
    } on Object {
      return StreamedResponse(
        Stream.fromIterable(
          [utf8.encode('Unable to connect to the server...')],
        ),
        504,
      );
    }
  }

  @override
  FutureOr<StreamedResponse> multipartRequest(
    String method,
    String urlSuffix,
    FutureOr<void> Function(MultipartRequest request) addToRequest,
  ) async {
    final fullUrl = '$apiServer$urlSuffix';
    final request = MultipartRequest(method, Uri.parse(fullUrl));
    await addToRequest(request);
    try {
      return await client.send(request);
    } on Object {
      return StreamedResponse(
        Stream.fromIterable(
          [utf8.encode('Unable to connect to the server...')],
        ),
        504,
      );
    }
  }
}


lib/src/response.dart
import 'dart:convert';

import 'package:event_db/event_db.dart';
import 'package:http/http.dart';

/// Adds functions to [StreamedResponse] to make it easier to work with.
extension FutureResponse on StreamedResponse {
  /// Returns the body as a future rather than as a stream. Can only be called
  /// once along with all the other body commands.
  Future<String> get body async => (await stream.toStringStream().toList())
      .reduce((value, element) => '$value$element');

  /// Returns the body as a future rather than as a stream. Can only be called
  /// once along with all the other body commands.
  ///
  /// This also automatically casts the [body] into a JSONObject map
  Future<Map<String, dynamic>> get bodyAsMap async =>
      json.decode(await body) as Map<String, dynamic>;

  /// Returns the body as a future rather than as a stream. Can only be called
  /// once along with all the other body commands.
  ///
  /// This also automatically transforms the [body] into the model [T]. The
  /// instance is generated by [supplier].
  Future<T> bodyAsModel<T extends BaseModel>(T Function() supplier) async =>
      supplier()..loadFromJsonString(await body);

  /// Returns the body as a future rather than as a stream. Can only be called
  /// once along with all the other body commands.
  ///
  /// This also automatically transforms the [body] into a list of model [T].
  /// The instance of each individual model is generated by [supplier].
  Future<List<T>> bodyAsModelList<T extends BaseModel>(
    T Function() supplier,
  ) async =>
      (json.decode(await body) as List)
          .map(
            (e) => supplier()..loadFromMap(e as Map<String, dynamic>),
          )
          .toList();
}


lib/src/repository.dart
import 'dart:async';

import 'package:event_api/event_api.dart';
import 'package:event_authentication/event_authentication.dart';
import 'package:event_bloc/event_bloc.dart';
import 'package:event_db/event_db.dart';
import 'package:http/http.dart';

const _encodedJWT = 'EncodedJWT';

/// Automatically manages JWTs while sending requests.
class APIRepository extends Repository {
  /// [database] is where the JWT will be saved.
  ///
  /// [requester] is what
  APIRepository({
    required this.database,
    required this.requester,
  });

  /// Where the [EncodedJWT] will be stored.
  final SpecificDatabase database;

  /// Handles sending requests.
  final APIRequester requester;

  EncodedJWT? _jwt;

  /// This is the JWT that will be attached to all requests.
  Future<EncodedJWT?> get jwt async {
    if (_jwt != null) {
      return _jwt;
    }
    return _jwt = await database.findModel(_encodedJWT);
  }

  /// Deletes the current jwt to perform a logout.
  Future<void> logout() async {
    final currentJwt = await jwt;
    if (currentJwt != null) {
      await database.deleteModel(currentJwt);
    }
    _jwt = null;
  }

  /// Creates a request using [method] to the [urlSuffix] endpoint with
  /// [requester].
  ///
  /// [addToRequest] is used for addings things like headers to the request
  /// before sending it.
  ///
  /// Automatically adds [jwt] as an Authorization header if it's available.
  Future<StreamedResponse> request(
    String method,
    String urlSuffix,
    FutureOr<void> Function(Request request) addToRequest,
  ) async {
    final response =
        await requester.request(method, urlSuffix, (request) async {
      final loadedJwt = await jwt;
      request.headers.authorization = loadedJwt?.token;
      addToRequest(request);
    });

    if (response.headers.authorization == null) {
      return response;
    }
    _jwt = EncodedJWT.fromToken(response.headers.authorization!)
      ..id = _encodedJWT;
    database.saveModel(_jwt!);
    return response;
  }

  /// Creates a multipart request using [method] to the [urlSuffix] endpoint
  /// with [requester].
  ///
  /// [addToRequest] is used for addings things like headers to the request
  /// before sending it.
  ///
  /// Automatically adds [jwt] as an Authorization header if it's available.
  Future<StreamedResponse> multipartRequest(
    String method,
    String urlSuffix,
    FutureOr<void> Function(MultipartRequest request) addToRequest,
  ) async {
    final response =
        await requester.multipartRequest(method, urlSuffix, (request) async {
      final loadedJwt = await jwt;
      request.headers.authorization = loadedJwt?.token;
      addToRequest(request);
    });

    if (response.headers.authorization == null) {
      return response;
    }
    _jwt = EncodedJWT.fromToken(response.headers.authorization!)
      ..id = _encodedJWT;
    database.saveModel(_jwt!);
    return response;
  }

  @override
  List<BlocEventListener<dynamic>> generateListeners(
    BlocEventChannel channel,
  ) =>
      [];
}

/// Manages sending requests to the server or other API.
abstract class APIRequester {
  /// The server to send requests to.
  String get apiServer;

  /// The expected client-side website to be used as the origin
  String get website;

  /// Creates a request using [method] to the [urlSuffix] endpoint.
  ///
  /// [addToRequest] is used for addings things like headers to the request
  /// before sending it.
  FutureOr<StreamedResponse> request(
    String method,
    String urlSuffix,
    void Function(Request request) addToRequest,
  );

  /// Creates a multipart request using [method] to the [urlSuffix] endpoint.
  ///
  /// [addToRequest] is used for addings things like headers to the request
  /// before sending it.
  FutureOr<StreamedResponse> multipartRequest(
    String method,
    String urlSuffix,
    void Function(MultipartRequest request) addToRequest,
  );
}


lib/src/database.dart
import 'dart:async';
import 'dart:convert';

import 'package:event_api/event_api.dart';
import 'package:event_bloc/event_bloc.dart';
import 'package:event_db/event_db.dart';

/// An implementation of [DatabaseRepository] that relies on an API Server or
/// [APIRepository] to manage the data.
class ApiDatabaseRepository extends DatabaseRepository {
  /// [apiRepository] is the api server where the [BaseModel]s will be stored
  ///
  /// [baseURL] is infixed to all requests. Should not end with /
  ApiDatabaseRepository({
    required this.apiRepository,
    required this.baseURL,
    required this.constructors,
  });

  /// [constructors] holds all of the [ModelConstructor]s that can be found in
  /// this repository.
  final Map<Type, ModelConstructor> constructors;

  /// The api server where the [BaseModel]s will be stored
  final APIRepository apiRepository;

  /// Infixed to all requests. Should not end with /
  final String baseURL;

  @override
  FutureOr<bool> deleteModel<T extends BaseModel>(
    String database,
    T model,
  ) async {
    assert(model.id != null, 'Cannot delete model with no specified id!');
    final response = await apiRepository.request(
      'DELETE',
      '$baseURL/$database/${model.id}',
      (request) => request.body = model.toJsonString(),
    );

    switch (response.statusCode) {
      case 200:
        return true;
      case 504:
        errorStream.add(
          DatabaseException(
            database: database,
            action: DatabaseAction.delete,
            error: DatabaseError.noDatabaseAccess,
          ),
        );
        return false;
      case 404:
      default:
        return false;
    }
  }

  @override
  FutureOr<Iterable<T>> findAllModelsOfType<T extends BaseModel>(
    String database,
    T Function() supplier,
  ) async {
    final response = await apiRepository.request(
      'GET',
      '$baseURL/$database',
      (request) => null,
    );

    switch (response.statusCode) {
      case 200:
        final body = await response.body;
        final list = json.decode(body) as List;
        return list
            .map((e) => supplier()..loadFromMap(e as Map<String, dynamic>));
      case 504:
        errorStream.add(
          DatabaseException(
            database: database,
            action: DatabaseAction.findAll,
            error: DatabaseError.noDatabaseAccess,
          ),
        );
        return [];
      case 404:
      default:
        return [];
    }
  }

  /// Creates an instance using one of the [constructors]
  ///
  /// Will throw an [ArgumentError] if one isn't present in [constructors]
  /// for [S]
  S getInstance<S>() {
    if (!constructors.containsKey(S)) {
      throw ArgumentError(
          '$S was not added to the constructors of this ApiDatabaseRepository.'
          ' Please ensure you add all constructors you want to use for each '
          'type you will use.');
    }
    return constructors[S]!() as S;
  }

  @override
  FutureOr<T?> findModel<T extends BaseModel>(
    String database,
    String key,
  ) async {
    final response = await apiRepository.request(
      'GET',
      '$baseURL/$database/$key',
      (request) => null,
    );

    switch (response.statusCode) {
      case 200:
        return response.bodyAsModel(getInstance<T>);
      case 504:
        errorStream.add(
          DatabaseException(
            database: database,
            action: DatabaseAction.findByKey,
            error: DatabaseError.noDatabaseAccess,
          ),
        );
        return null;
      case 404:
      default:
        return null;
    }
  }

  @override
  List<BlocEventListener<dynamic>> generateListeners(
    BlocEventChannel channel,
  ) =>
      [];

  @override
  FutureOr<T> saveModel<T extends BaseModel>(
    String database,
    T model,
  ) async {
    final shouldUpdate = model.id != null;
    final response = await (shouldUpdate
        ? apiRepository.request(
            'PUT',
            '$baseURL/$database/${model.id}',
            (request) => request.body = model.toJsonString(),
          )
        : apiRepository.request(
            'POST',
            '$baseURL/$database',
            (request) => request.body = model.toJsonString(),
          ));

    switch (response.statusCode) {
      case 200:
        return model..id = await response.body;
      case 504:
        errorStream.add(
          DatabaseException(
            database: database,
            action: DatabaseAction.save,
            error: DatabaseError.noDatabaseAccess,
          ),
        );
        return model;
      case 404:
      default:
        return model;
    }
  }
}
